Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -280,43 +290,48 @@ class HTTP1ConnectionProvider {
         self.state.assertInvariants()
     }
 
-    private func execute(_ action: Action) {
+    private func execute(_ action: Action, logger: Logger) {
         switch action {
         case .lease(let connection, let waiter):
             // if connection is became inactive, we create a new one.
             connection.cancelIdleTimeout().whenComplete { _ in
                 if connection.isActiveEstimation {
+                    logger.debug("reusing active connection \(connection)")
```

Comment:
Not too important for logs, that meant to me disabled by default, but when you have logs aggregating system, it's useful to only have static log messages with parameters in metadata. 
e.g.
```suggestion
                    logger.debug("reusing active connection", metadata: ["connection":"\(connection)"])
```
Also, looking at `class HTTP1ConnectionProvider`, I wonder if we're writing too much information here? I'd expect something like (connectionId, connectionState).

Reply:
Thanks @avolokhov! So I'm personally happy doing this change but I'd like to get the opinion of a few folks. Let me first be very clear what you're proposing so we're all on the same page.

Anton proposes that we make _all_ log messages _fully static_, ie. no string interpolation in them. As an example, instead of

    logger.info("successfully updated \(colour.name)'s hex representation to \(colour.htmlHexRepresentation), DB UPDATE modified \(modifiedRows) rows.")

which would log something like

    2020-05-21T18:56:16+0100 info: successfully updated red's hex representation to #ff0000, DB update modified 1 rows.

which uses three string interpolations, you'd do

    logger.info("successfully updated colour",
                metadata: ["colour": "\(colour.name)",
                           "hex-colour": "\(colour.htmlHexRepresentation)",
                           "modified-rows": "\(modifiedRows)"])

logging something like

    2020-05-21T18:56:16+0100 info: colour=red hex-color=#ff0000 modified-rows=1 successfully updated colour

Obviously very much dependent on your actual logger. A more real-world production scenario where the log message is more or less just another metadata field, it could look like

    {"timestamp: "2020-05-21T18:56:16+0100", level: "info", message: "successfully updated red's hex representation to #ff0000, DB update modified 1 rows."}

vs.

    {"timestamp: "2020-05-21T18:56:16+0100", level: "info", message: "successfully updated colour", colour: "red", "hex-colour": "#ff0000", "modified-rows": 1}
    


In other words: The log _message_ would always just tell you the objective of what has/will happen. The the metadata would give you information about the parameters specific to this invocation.

Why is that useful? In serious production, most people will use log aggregation systems and often you'll filter on something like `system="acme-001" log-message="successfully updated colour"` which works really well if the log messages are always static.

Especially interspersed messages like my example above

    successfully updated red's hex representation to #ff0000, DB update modified 1 rows.

are very awkward to search for, you'd either need regexes or globs to do search for all of those messages (`successfully updated *'s hex representation to*, DB update modified * rows.`). That could be just awkward (the debugging engineer needs to "reverse engineer" which bits are the variables and what's the log message) or impossible (globs/regexes not supported in you log aggregator (unlikely)). Most likely I'd think it's just slower to grep for globs/regexes and more awkward.

What about downsides? I'd say as a potential downside, it's probably slightly harder for humans to read what's going on. The real downside however may happen if we have users that _do not_ log the metadata at all... I'm not sure how likely this is but it's a problem.

One other solution that I see would be to put all the variable data at the _end_ of a log message. Ie. something like

        logger.info("successfully updated colour hex representation. name=\(colour.name),  hex representation=\(colour.htmlHexRepresentation), modified rows=\(modifiedRows) rows.")

That _probably_ allows for fast grepping (as you just need the prefix) but does feel wrong for me because the user pretty much does metadata but embedded into the log message.

I hope this makes some some, please ask/correct me if there questions/concerns/misinterpretations.

CC @tomerd / @avolokhov / @tanner0101 / @adam-fowler / @timburks / @glbrntt / @Lukasa / @tachyonics / @artemredkin

Reply:
Putting all variables at the end of log line will help with manual searches, but not automated reports - there's no way to automate random truncation for random log lines. As I mentioned, it's not very important here as you don't normally want to automate debug logs aggregation.

Reply:
Thanks @avolokhov , that's a strong argument I think. I'm in favour of changing it to what Anton proposes.

Reply:
> Anton proposes that we make all log messages fully static, ie. no string interpolation in them. [...]

I'm totally in favor of that if we can get others on board.

I'm doing that in my lib (more and more) and would strongly recommend doing that if we can others on board (usually people hate on me when I propose this so was not pushing it onto folks here 😉).

This then expands onto proper tooling where one wants to filter by tags (even just the keys sometimes is useful), and then driling down to limit by a specific value. I also have some "debugging" logger impls which print the metadata more human readable as:

```
[info] thing happened
// key: VeryLong(
//  value: Keys (
//    ... 
//  )
// key2: ... 
```

etc which I find invaluable in debugging/ development time with _tons_ of logs :-)
Such formatters are not really easily doable with in-line text, so another great reason to do the metadata only style.

---

Then one can it another level, since the metadata gets quite repetetive at times.
For example if one always logs request but also trace id, but if available some other thing...

building the metadata becomes:

```swift
var metadata = ... 
if let traceid ... {
  metadata[...] = traceid
}
metadata["request-id"] = requestID
```

etc.

So this can quickly grow quite big, and repetetive if one logs the same thing in 10 places.

What I've been doing for this is to do metadata functions on objects, so `ThingThatHasThings.metadata(anything extra)` can be used to encapsulate that metadata builderplate:

```swift
log.info("hello", metadata: ThingThatHasThings.metadata(somethingSpecial: extra)
```

so that helps avoid the boilerplate :)

Johannes in this PR did some interesting encapsulating this in `logger.someSpecific.info()` which is also interesting... I guess the goal is the same really, hiding the boilerplate away.

All in all, +1, I think it's definitely worth investing in more metadata style logging, rather than plain boring text which is easy to break if anyone has scripts grepping logs and issuing metrics or alters from them etc.

Reply:
Another thought: for what it's worth, if we adopt metadata-first style in our libs then it'll be even more natural for the context propagation / tracing how it'll fit in 🥇 Since context propagation will be via `context.log` where it populates metadata, so it'll feel very "natural" if all other data is exposed the same way :-)

Reply:
I'm definitely `+1` on the metadata approach. Looking at it from the perspective of using the logs to find performance hick-ups because of closed connections (example debugging DynamoDB), I'd hope to just filter for a domain/ip and see the connection history of all connections towards that domain/ip.

Reply:
`+1` on static messages here.

@weissi perhaps this is something the SSWG should provide guidance on.

Reply:
This makes total sense to me. Will ensure we do the same when implementing logging in aws-sdk-swift.

Reply:
Just a couple of points here -

I'd be careful of overloading the meaning of the logging metadata. Metadata is often (currently mainly maybe?) used to record the *context* of the log message - request ids, trace ids etc.

My concern would be that a proposal like this is potentially confusing the meaning of logging metadata. This is probably not an issue if it is just what one library decides to do but could be an issue if it is something we want the broader ecosystem to adopt.

Not all logging implementations may want to treat the context of a log message the same as the *parameters* of a log message. It may be worth adding overrides of the log methods that accept a static log message + parameters for clarity and also to allow logging implementations to handle these parameters either similarly or differently to metadata.

On the broader point of log aggregating systems, they are definitely useful and we should make sure they are supported as easily as possible. However, particularly for quantities that we know people are likely wanting to track - such as creating vs reusing connections, emitting metrics seems like a more appropriate mechanism to provide by default rather than relying on log aggregation.

This might be a discussion for another PR but this is what we do for SmokeAWS/SmokeHTTP with the ability to disable emitting metrics based on passed in configuration. Ideally this is a mechanism we could extend to any metrics emitted at this lower layer-

1. https://github.com/amzn/smoke-aws/blob/3dbe5a0ff90e30be829d6e126b25523fd57448a9/Sources/SmokeAWSCore/SmokeAWSClientReportingConfiguration.swift#L20
2. https://github.com/amzn/smoke-aws/blob/3dbe5a0ff90e30be829d6e126b25523fd57448a9/Sources/SmokeAWSCore/SmokeAWSOperationReporting.swift#L92-L100
3. https://github.com/amzn/smoke-http/blob/ae8003b3605c0b368925558254055222237b1da4/Sources/SmokeHTTPClient/HTTPOperationsClient%20%2BexecuteSyncRetriableWithOutput.swift#L82

Reply:
+1 on static messages, logging aggregates are useful. Also, in my main project we can specify two messages, one with normal "static" message, and one "human-readable" template, ie:
```swift
logger.info("reusing connection",
        humanReadable: "reusing connection with id {id}",
        metadata: ["id": "\(connection.id)"])
```
That way one can have two loggers, one (structured) can write static strings, and a second one will write interpolated string.

Reply:
@tachyonics Thank you for commenting! These are also really good points. Using metrics for the "core metrics" like connection reuse sounds like the right approach to me. I'm still leaning towards static messages (in addition to metrics) for more detailed stats.

For example: In AHC, it's not a binary decision: Do we reuse a connection or do we not. There are a couple of steps involved before AHC really commits to reusing one connection. For example, as one of the very last things, AHC probes the `Channel` for its `.isActive` property and if that just flipped to `false` it'll abandon that `Channel` and establish a fresh one. Of course, this is racy and will not always prevent a request from failing but I guess it's just another chance. Of course, we could have a "got-a-connection-but-abandoned-because-inactive-just-before-sending" metric to capture those but I think that may be a bit over the top. But I could imagine that if debugging some transiently failing HTTP requests it may still be useful to filter these "we tried a connection but then aborted in the last second" kind of messages. To do that, static messages could still be useful? WDYT? And @avolokhov ?

@artemredkin how do you communicate the 'human' vs. 'static' message to the `Logger`? Is the 'human' one the actual message and the 'static' one lives in the metadata but your `LogHandler` knows the "special" metadata field and uses that as the message when sending it to the log aggregator?

Reply:
@weissi That makes sense; core metrics to track the general health of the system, allowing it to alarm if something is looking really wrong and searchable logs to dig in and debug a specific problem.

Reply:
@tachyonics that's what I was _thinking_ but on the decision whether to still use static messages everywhere (or not, or have static prefixes, or ...) I was planning to stand on the shoulders of you giants who have more operational expertise than I do.

Reply:
As noted by @fabianfett, when debugging we usually try to trace some particular connection/uuid. I'd definitely favour an approach that allows `SELECT * FROM logs WHERE <ahc_connection_id/my_request_uuid>=$0` out of the box.
Interesting point re context vs parameters. I see why we may want to separate them, but it may be difficult to draw the line between the two. For example, `SELECT * FROM logs WHERE connection_id=$0 AND action=lease` is a completely legit query and it may make sense during debugging in certain circumstances. Or I can picture myself running something like `SELECT request_uuid FROM logs WHERE connection_id=$0 AND 'ahc_request_time_ms > 100 LIMIT 1' when trying to investigate slow requests.

Reply:
@avolokhov I mean we _could_ prefix them. `ahc-context-request-id` and `ahc-parameter-connection`.

Ideally, we'd try to keep the strings in 15 or fewer characters because then, they can be 'inline Strings' and don't ever require a heap allocation. That's not really relevant for constant strings [they point to a shared global] but if we were to construct the key names say [using `extension Logger`]

    logger.info("connection established", parameters: ["connection": connection], context: ["request-id": requestID])

which would then construct the labels and invoke

   logger.info("connection established", metadata: ["ahc-parameter-connection": connection, "ahc-context-request-id": requestID]

then it may matter. So maybe we shorten them to `ahc-c:` (for context) and `ahc-p:` (for parameters) that would leave us 9 characters to still be an inline String :P. Maybe this is going too far though?

Reply:
Yeah, maybe too much. And since the line between the two may be blurry at times, it may produce confusion.